### PR TITLE
Consistency Grid - Second Try

### LINKED ISSUE #423: Consistency check for extents and offsets
We need to check whether all extents and offsets are of the grid type that the user promises us.

We already know which grid type we have, so we need to traverse the AST, check all offsets and make sure that they can be casted to that grid type.

This consistency check should be unit tested!

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

This is a second try to impose a grid type check (firs try is PR #583). This time, the grid type checking was made part of the validator. The grid type validation is now done using the same principle as the location type checking. 

### Resolves / Enhances

Fixes #423 

### Testing

A negative test, i.e. ensuring that an IIR with mixed grid types is rejected has been introduced. The other way around, i.e. that there are no false positives is implied by no  failing existing tests.

### Note

There is currently no infrastructure to launch all consistency tests, or to launch consistency tests after each pass. The consistency test at hand was simply copy pasted to after each call to the location consistency test. Infrastructure w.r.t this should be added. This will be formulated as a follow up task to keep the current tasks within a reasonable time frame.

### CODE REVIEW

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/src/dawn/Validator/GridTypeChecker.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        }
+      }
+    }
+  }
+
+  for(const auto& doMethodPtr : iterateIIROver<iir::DoMethod>(iir)) {
+    GridTypeChecker::TypeCheckerImpl Impl(iir.getGridType());
+    const std::shared_ptr<iir::BlockStmt>& ast =
+        std::make_shared<iir::BlockStmt>(doMethodPtr->getAST());
+    ast->accept(Impl);
```
Comment (twicki) [inline dawn/src/dawn/Validator/GridTypeChecker.cpp]: impl is not a very specific name

File: dawn/src/dawn/Validator/GridTypeChecker.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      return false;
+    }
+  }
+  return true;
+}
+bool GridTypeChecker::checkGridTypeConsistency(const dawn::SIR& sir) {
+  // check type consistency of stencil functions
+  for(auto const& stenFunIt : sir.StencilFunctions) {
+    for(const auto& astIt : stenFunIt->Asts) {
+      GridTypeChecker::TypeCheckerImpl Impl(sir.GridType);
```
Comment (twicki) [inline dawn/src/dawn/Validator/GridTypeChecker.cpp]: again, could be moved out?

File: dawn/src/dawn/Optimizer/PassManager.cpp [hunk 3]
```diff
@@ -69,6 +80,6 @@ bool PassManager::runPassOnStencilInstantiation(
   passCounter_[pass->getName()]++;
   DAWN_LOG(INFO) << "Done with " << pass->getName() << " : Success";
   return true;
-}
+} // namespace dawn
```
Comment (twicki) [inline dawn/src/dawn/Optimizer/PassManager.cpp]: probably in here due to mismatching parentheses at some point

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**